src/textual/app.py

```diff
@@ -1822,23 +1824,66 @@ async def _on_resize(self, event: events.Resize) -> None:
         await self.screen.post_message(event)
 
     async def _on_remove(self, event: events.Remove) -> None:
-        widget = event.widget
-        parent = widget.parent
+        """Handle a remove event.
 
-        remove_widgets = widget.walk_children(
-            Widget, with_self=True, method="depth", reverse=True
-        )
+        Args:
+            event (events.Remove): The remove event.
+        """
 
-        if self.screen.focused in remove_widgets:
+        # We've been given a list of widgets to remove, but removing those
+        # will also result in other (descendent) widgets being removed. So
+        # to start with let's get a list of everything that's not going to
+        # be in the DOM by the time we've finished. Note that, at this
+        # point, it's entirely possible that there will be duplicates.
+        everything_to_remove: list[Widget] = []
+        for widget in event.widgets:
+            everything_to_remove.extend(
+                widget.walk_children(
+                    Widget, with_self=True, method="depth", reverse=True
+                )
+            )
+
+        # Next up, let's quickly create a deduped collection of things to
+        # remove and ensure that, if one of them is the focused widget,
+        # focus gets moved to somewhere else.
+        dedupe_to_remove = set(everything_to_remove)
+        if self.screen.focused in dedupe_to_remove:
             self.screen._reset_focus(
                 self.screen.focused,
-                [to_remove for to_remove in remove_widgets if to_remove.can_focus],
+                [to_remove for to_remove in dedupe_to_remove if to_remove.can_focus],
             )
 
-        await self._prune_node(widget)
-
-        if parent is not None:
-            parent.refresh(layout=True)
+        # Next, we want to go through the full set of everything to remove
+        # and reduce it to the minimal set of things that need removing. In
+        # other words, for any given branch that is going to have things
+        # removed, we need to just keep the topmost node in the DOM that is
+        # a candidate for removal. Note that we retain the ordering of the
+        # original walk at the top as it's depth-first and so should be the
+        # order we want to go in from here on.
+        pruned_remove: list[Widget] = [
+            widget
+            for widget in everything_to_remove
+            if not any(
```

Comment:
I think we can replace the if condition with a set operation:

```python
isdisjoint = everything_to_remove
pruned_remove = [
    widget for widget in everything_to_remove
    if isdisjoint(widget.ancestors)
]
```

Reply:
Like this? https://github.com/Textualize/textual/pull/1147/commits/3cb00f93d5e2b000f2fa19951b5338ac0ee59180